src/ppx/application.cpp

```diff
@@ -1273,6 +1213,282 @@ int Application::Run(int argc, char** argv)
+    FrameRender();
+
+    // Take screenshot if this is the requested frame.
+    if (mFrameCount == static_cast<uint64_t>(mStandardOpts.pScreenshotFrameNumber->GetValue())) {
+        TakeScreenshot();
+    }
+
+    // Frame end general metrics data, used for recorded metrics, display, screenshots, and pacing.
+    double nowMs       = mTimer.MillisSinceStart();
+    mFrameCount        = mFrameCount + 1;
+    mPreviousFrameTime = static_cast<float>(nowMs) - mFrameStartTime;
+
+    // Keep a rolling window of frame times to calculate stats, if requested.
+    if (mStandardOpts.pStatsFrameWindow->GetValue() > 0) {
+        mFrameTimesMs.push_back(mPreviousFrameTime);
+        if (mFrameTimesMs.size() > mStandardOpts.pStatsFrameWindow->GetValue()) {
+            mFrameTimesMs.pop_front();
+        }
+        float totalFrameTimeMs = std::accumulate(mFrameTimesMs.begin(), mFrameTimesMs.end(), 0.f);
+        mAverageFPS            = mFrameTimesMs.size() / (totalFrameTimeMs / 1000.f);
+        mAverageFrameTime      = totalFrameTimeMs / mFrameTimesMs.size();
+    }
+    else {
+        mAverageFPS       = static_cast<float>(mFrameCount / (nowMs / 1000.f));
+        mAverageFrameTime = static_cast<float>(nowMs / mFrameCount);
+    }
+
+    // Update the metrics. This can be used for both recorded AND displayed metrics,
+    // and therefore should always be called.
+    DispatchUpdateMetrics();
+
+    // Pace frames - if needed
+    if (mSettings.grfx.pacedFrameRate > 0) {
+        if (mFrameCount > 0) {
+            double currentTime  = nowMs / 1000.f;
+            double pacedFPS     = 1.0 / static_cast<double>(mSettings.grfx.pacedFrameRate);
+            double expectedTime = mFirstFrameTime + (mFrameCount * pacedFPS);
+            double diff         = expectedTime - currentTime;
+            if (diff > 0) {
+                Timer::SleepSeconds(diff);
+            }
+        }
+        else {
+            mFirstFrameTime = nowMs / 1000.f;
+        }
+    }
+    // If we reach the maximum number of frames allowed
+    if ((mStandardOpts.pFrameCount->GetValue() > 0 && mFrameCount >= mStandardOpts.pFrameCount->GetValue()) ||
+        (nowMs / 1000.f) > mRunTimeSeconds) {
+        Quit();
+    }
+    return;
+}
+
+int Application::Run(int argc, char** argv)
+{
+    // Only allow one instance of Application. Since we can't stop
+    // the app in the ctor - stop it here.
+    //
+    if (this != sApplicationInstance) {
+        return false;
+    }
+
+    // Parse args.
+    if (Failed(mCommandLineParser.Parse(argc, const_cast<const char**>(argv)))) {
+        PPX_ASSERT_MSG(false, "Unable to parse command line arguments");
+        return EXIT_FAILURE;
+    }
+
+    // Call config.
+    // Put this early because it might disable the display.
+    DispatchConfig();
+
+    // Knobs need to be set up before commandline parsing.
+    // This has dependency on DispatchConfig() where standard knob default values are set
+    DispatchInitKnobs();
```

Comment:
Same here?

```cpp
InitStandardKnobs();
// Let the application initialize their own knobs.
InitKnobs();
```

Reply:
i'd prefer keep them in a function so that knob initializations are in the same func for clarity (also being consistent with other part of the code in this file.)